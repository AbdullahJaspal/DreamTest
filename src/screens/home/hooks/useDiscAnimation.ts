import {useEffect, useCallback} from 'react';
import {
  useSharedValue,
  withRepeat,
  withTiming,
  Easing,
  useAnimatedStyle,
} from 'react-native-reanimated';

const useDiscAnimation = () => {
  const discAnimatedValue = useSharedValue(0);

  const startAnimation = useCallback(() => {
    discAnimatedValue.value = withRepeat(
      withTiming(360, {
        duration: 5000,
        easing: Easing.linear,
      }),
      -1,
      false,
    );
  }, [discAnimatedValue]);

  useEffect(() => {
    startAnimation();
  }, [startAnimation]);

  const discAnimatedStyle = useAnimatedStyle(() => {
    return {
      transform: [
        {
          rotate: `${discAnimatedValue.value}deg`,
        },
      ],
    };
  });

  return discAnimatedStyle;
};

export default useDiscAnimation;
