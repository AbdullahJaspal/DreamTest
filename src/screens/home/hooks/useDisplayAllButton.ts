import React from 'react';

const useDisplayAllButton = () => {};

export default useDisplayAllButton;
