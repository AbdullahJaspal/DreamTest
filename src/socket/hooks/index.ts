export {default as useSocket} from '../SocketProvider';
export {default as useChatSocket} from './useChatSocket';
export {default as usePostSocket} from './usePostSocket';
export {default as useCommentSocket} from './useCommentSocket';
