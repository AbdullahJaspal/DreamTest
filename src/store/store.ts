import {configureStore} from '@reduxjs/toolkit';

import {rootReducer} from './rootReducer';
const preloadedState = loadFromLocalStorage();

export const store = configureStore({
  reducer: rootReducer,
  preloadedState,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: false,
      immutableCheck: false,
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

store.subscribe(() => {
  saveToLocalStorage(store.getState());
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

function loadFromLocalStorage() {
  try {
    const serializedState = localStorage.getItem('appState');
    if (serializedState === null) {
      return undefined;
    }
    return JSON.parse(serializedState);
  } catch (e) {
    console.warn('Error loading state from localStorage:', e);
    return undefined;
  }
}

function saveToLocalStorage(state: RootState) {
  try {
    const stateToPersist = {
      themeColor: state.themeColor,
      my_data: {
        isLogin: state.my_data.isLogin,
        my_profile_data: state.my_data.my_profile_data,
      },
    };
    const serializedState = JSON.stringify(stateToPersist);
    localStorage.setItem('appState', serializedState);
  } catch (e) {
    console.warn('Error saving state to localStorage:', e);
  }
}
