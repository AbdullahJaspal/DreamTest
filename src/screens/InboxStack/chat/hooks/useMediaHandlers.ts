import {useCallback} from 'react';
import Toast from 'react-native-simple-toast';
import {useTranslation} from 'react-i18next';
import {useSocket} from '../../../../socket/SocketProvider';

import {
  MessageStatus,
  ExtendedMessage,
  generateUniqueId,
  uploadMedia,
} from '../../../../utils/chatUtils';

// Redux imports
import {useAppDispatch, useAppSelector} from '../../../../store/hooks';
import {
  appendMessage,
  updateMessageStatus,
  setUploading,
} from '../../../../store/slices/content/messageSlice';
import {makeSelectIsUploadingForRoom} from '../../../../store/selectors';

// Define proper interfaces for user data
interface UserData {
  id: string;
  nickname: string;
  profile_pic: string;
  auth_token?: string;
}

// Media data interface
interface MediaData {
  uri: string;
  caption?: string;
  duration?: string;
  fileName?: string;
  mimeType?: string;
}

export const useMediaHandlers = (
  roomId: string,
  myData: UserData,
  userData: UserData,
) => {
  const {t} = useTranslation();
  const {isConnected, emit} = useSocket();
  const dispatch = useAppDispatch();

  const selectIsUploadingForCurrentRoom = makeSelectIsUploadingForRoom();
  const isUploading = useAppSelector(state =>
    selectIsUploadingForCurrentRoom(state, roomId),
  );

  const prepareUserData = useCallback(
    () => ({
      _id: myData.id,
      name: myData.nickname,
      avatar: myData.profile_pic,
    }),
    [myData],
  );

  // Helper to update message status
  const updateMessageStatusWithProgress = useCallback(
    (messageId: string, status: MessageStatus, uploadProgress = 0) => {
      dispatch(
        updateMessageStatus({
          roomId,
          messageId,
          status,
          uploadProgress,
        }),
      );
    },
    [dispatch, roomId],
  );

  // Upload media file to server
  const uploadMediaToServer = useCallback(
    async (
      mediaUri: string,
      mediaType: string,
      fileName = null,
      mimeType = null,
    ) => {
      try {
        return await uploadMedia(
          mediaUri,
          mediaType,
          myData.auth_token,
          fileName,
          mimeType,
        );
      } catch (error) {
        console.error(`Error uploading ${mediaType}:`, error);
        throw error;
      }
    },
    [myData.auth_token],
  );

  // Generic function to handle all media types
  const uploadAndSendMedia = useCallback(
    async (
      mediaData: MediaData,
      mediaType: string,
      replyToMessage?: ExtendedMessage | null,
      existingId?: string,
    ) => {
      const messageId = existingId || generateUniqueId();

      // Create placeholder message to show in UI immediately
      const placeholderMessage: ExtendedMessage = {
        _id: messageId,
        senderId: myData.id,
        receiverId: userData.id,
        text: mediaData.caption || '',
        type: mediaType,
        image: mediaType === 'image' ? mediaData.uri : '',
        video: mediaType === 'video' ? mediaData.uri : '',
        audio: mediaType === 'audio' ? mediaData.uri : '',
        document: mediaType === 'document' ? mediaData.uri : '',
        parentMessageId: replyToMessage ? replyToMessage._id : '',
        replyToMessage: replyToMessage || '',
        duration: mediaData.duration,
        createdAt: new Date(),
        user: prepareUserData(),
        status: MessageStatus.SENDING,
        uploadProgress: 0,
        roomId: roomId,
      };

      // Add message to Redux
      if (existingId) {
        updateMessageStatusWithProgress(existingId, MessageStatus.SENDING, 0);
      } else {
        // We need to flag this as a local-only message before the upload is complete
        // This will help us identify it if we get a duplicate from the server
        const localMessage = {
          ...placeholderMessage,
          isLocalPlaceholder: true,
        };
        dispatch(appendMessage({roomId, message: localMessage}));
      }

      try {
        dispatch(setUploading({roomId, isUploading: true}));

        // Upload the media file
        const mediaUrl = await uploadMediaToServer(
          mediaData.uri,
          mediaType,
          mediaData.fileName,
          mediaData.mimeType,
        );

        if (!mediaUrl) {
          throw new Error('Upload failed');
        }

        // Create final message with remote media URL
        const finalMessage: ExtendedMessage = {
          ...placeholderMessage,
          image: mediaType === 'image' ? mediaUrl : '',
          video: mediaType === 'video' ? mediaUrl : '',
          audio: mediaType === 'audio' ? mediaUrl : '',
          document: mediaType === 'document' ? mediaUrl : '',
          status: MessageStatus.SENT,
          uploadProgress: 100,
          isLocalPlaceholder: false,
        };

        // Update message in Redux
        updateMessageStatusWithProgress(messageId, MessageStatus.SENT, 100);

        // Send message via socket
        if (isConnected) {
          emit('customEvent', [finalMessage]);
        } else {
          throw new Error('No connection');
        }

        return true;
      } catch (error) {
        console.error(`Error sending ${mediaType}:`, error);
        updateMessageStatusWithProgress(messageId, MessageStatus.FAILED);

        let errorMessage = `Failed to send ${mediaType}`;
        if (error instanceof Error) {
          if (error.message.includes('No connection')) {
            errorMessage = t('No internet connection. Please try again later.');
          }
        }

        Toast.show(errorMessage, Toast.SHORT);
        return false;
      } finally {
        dispatch(setUploading({roomId, isUploading: false}));
      }
    },
    [
      myData.id,
      userData.id,
      prepareUserData,
      uploadMediaToServer,
      isConnected,
      roomId,
      emit,
      updateMessageStatusWithProgress,
      dispatch,
      t,
    ],
  );

  // Generic handler for all media types
  const handleSendMedia = useCallback(
    async (
      mediaData: MediaData,
      mediaType: string,
      replyToMessage?: ExtendedMessage | null,
    ) => {
      return await uploadAndSendMedia(mediaData, mediaType, replyToMessage);
    },
    [uploadAndSendMedia],
  );

  // Simplified handlers for specific media types, using the generic handler
  const handleSendImage = useCallback(
    async (imageData: any, replyToMessage?: ExtendedMessage | null) => {
      return await handleSendMedia(
        {
          uri: imageData.uri,
          caption: imageData.caption,
          duration: imageData.duration,
        },
        'image',
        replyToMessage,
      );
    },
    [handleSendMedia],
  );

  const handleSendVideo = useCallback(
    async (videoData: any, replyToMessage?: ExtendedMessage | null) => {
      return await handleSendMedia(
        {
          uri: videoData.uri,
          caption: videoData.caption,
          duration: videoData.duration,
        },
        'video',
        replyToMessage,
      );
    },
    [handleSendMedia],
  );

  const handleSendDocument = useCallback(
    async (
      documentUri: string,
      documentName?: string,
      documentType?: string,
    ) => {
      return await handleSendMedia(
        {
          uri: documentUri,
          fileName: documentName,
          mimeType: documentType,
        },
        'document',
      );
    },
    [handleSendMedia],
  );

  const handleSendAudio = useCallback(
    async (audioUri: string, audioName?: string) => {
      return await handleSendMedia(
        {
          uri: audioUri,
          fileName: audioName,
        },
        'audio',
      );
    },
    [handleSendMedia],
  );

  // Handle sending contact information
  const handleSendContact = useCallback(
    (contact: any) => {
      if (!contact) return false;

      const contactData =
        typeof contact === 'object'
          ? contact
          : {displayName: 'Unknown', phoneNumbers: []};

      const phoneNumber =
        contactData.phoneNumbers && contactData.phoneNumbers.length > 0
          ? contactData.phoneNumbers[0].number
          : '';

      const contactMessage: ExtendedMessage = {
        _id: generateUniqueId(),
        senderId: myData.id,
        receiverId: userData.id,
        text: `Contact: ${contactData.displayName}\nPhone: ${phoneNumber}`,
        type: 'contact',
        contactName: contactData.displayName,
        contactPhone: phoneNumber,
        image: '',
        video: '',
        document: '',
        audio: '',
        roomId: roomId,
        createdAt: new Date(),
        user: prepareUserData(),
        status: isConnected ? MessageStatus.SENDING : MessageStatus.FAILED,
      };

      // Add to Redux
      dispatch(appendMessage({roomId, message: contactMessage}));

      // Send via socket
      if (isConnected) {
        emit('customEvent', [contactMessage]);

        // Update status to sent after a delay
        setTimeout(() => {
          updateMessageStatusWithProgress(
            contactMessage._id,
            MessageStatus.SENT,
          );
        }, 500);

        return true;
      } else {
        updateMessageStatusWithProgress(
          contactMessage._id,
          MessageStatus.FAILED,
        );
        return false;
      }
    },
    [
      myData.id,
      userData.id,
      roomId,
      prepareUserData,
      isConnected,
      emit,
      updateMessageStatusWithProgress,
      dispatch,
    ],
  );

  // Handle sending location
  const handleLocationSelect = useCallback(
    (location: any) => {
      if (!location) return;

      const locationMessage: ExtendedMessage = {
        _id: generateUniqueId(),
        senderId: myData.id,
        receiverId: userData.id,
        text: location.name || t('Shared location'),
        type: 'location',
        location: {
          latitude: location.latitude,
          longitude: location.longitude,
          name: location.name || t('Shared location'),
        },
        image: '',
        video: '',
        document: '',
        audio: '',
        roomId: roomId,
        createdAt: new Date(),
        user: prepareUserData(),
        status: isConnected ? MessageStatus.SENDING : MessageStatus.FAILED,
      };

      // Add to Redux
      dispatch(appendMessage({roomId, message: locationMessage}));

      // Send via socket
      if (isConnected) {
        emit('customEvent', [locationMessage]);

        // Update status to sent after a delay
        setTimeout(() => {
          updateMessageStatusWithProgress(
            locationMessage._id,
            MessageStatus.SENT,
          );
        }, 500);
      } else {
        updateMessageStatusWithProgress(
          locationMessage._id,
          MessageStatus.FAILED,
        );
      }
    },
    [
      myData.id,
      userData.id,
      roomId,
      prepareUserData,
      isConnected,
      emit,
      updateMessageStatusWithProgress,
      dispatch,
      t,
    ],
  );

  return {
    isUploading,
    handleSendMedia,
    handleSendImage,
    handleSendVideo,
    handleSendAudio,
    handleSendDocument,
    handleSendContact,
    handleLocationSelect,
  };
};
