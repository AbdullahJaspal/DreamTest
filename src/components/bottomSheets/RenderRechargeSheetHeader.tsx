import {
  Dimensions,
  Image,
  Pressable,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import React, {useState} from 'react';
import {icons} from '../../assets/icons';
import {useAppSelector} from '../../store/hooks';
import InfoModal from '../../components/infoButtonModal';
import {selectMyProfileData} from '../../store/selectors';

const {width} = Dimensions.get('screen');

const RenderRechargeSheetHeader = () => {
  const my_data = useAppSelector(selectMyProfileData);
  const [showInfo, setShowInfo] = useState(false);

  const faqData = [
    {
      question: 'How to Recharge Diamonds in Dream?',
      answer: `
1. Recharge Through Your Wallet:
   - Open the Dream App and go to the Wallet section from your profile.
   - Tap on “Recharge Diamonds” and browse the available diamond packages.
   - Select your preferred package and complete the transaction using your linked payment method.

2. Recharge While Sending Gifts:
   - Tap on the Gift Icon while watching a video.
   - If you don’t have enough diamonds, select “Recharge Diamonds” at the bottom of the screen.
   - Choose a diamond package and follow the payment process.

**Important Note:**
All purchases must be completed through the payment methods linked to your Google Play or App Store account. Make sure your payment method is valid and up to date for a smooth experience.
        `,
    },
    {
      question: 'Why Is My Recharge Restricted?',
      answer: `
To ensure account security, Dream temporarily disables recharges if unusual activity or suspicious transactions are detected. For assistance in resolving this issue, reach out to Customer Support through the app.
        `,
    },
    {
      question: 'Where Are Diamond Purchases Available?',
      answer: `
Currently, purchasing diamonds is only available in selected countries and regions. Additionally, users must meet the minimum age requirement:

- 18 years old in most countries.
- 19 years old in Korea.

Dream does not support third-party payment services. All purchases should be made directly through the official Dream App or website.
        `,
    },
    {
      question: 'What to Do If Payment Fails?',
      answer: `
If your payment cannot be processed, try the following steps:

1. Ensure your payment method is correctly linked to your Google Play or App Store account.
2. Avoid using a VPN during the transaction process.
3. Update your Dream App to the latest version.
4. Check your internet connection to ensure it is stable.

For persistent issues, visit the following support pages for additional help:
- Google Play Help Center
- Apple Support

Alternatively, report the issue directly through the Help Center in the Dream App.
        `,
    },
    {
      question: 'What If I Didn’t Receive My Diamonds?',
      answer: `
If you’ve successfully completed a payment but the diamonds aren’t appearing in your wallet:

1. Restart the Dream App and refresh the Wallet section.
2. Check your purchase receipt in the confirmation email sent by Google Play or App Store.
3. If the issue persists, take a screenshot of the receipt showing:
   - Order ID
   - Transaction Amount
   - Payment Time

Submit the screenshot via the Help Center in the Dream App for a quick resolution.
        `,
    },
    {
      question: 'Need Further Assistance?',
      answer: `
If you’re experiencing issues or have more questions about diamond purchases, reach out to Dream Customer Support through the Help Center in the app. We’re here to help!
        `,
    },
  ];

  return (
    <>
      <View style={styles.upper_container}>
        <Text style={styles.recharge_text}>Recharge</Text>

        <Pressable
          style={styles.question_mark_view}
          onPress={() => setShowInfo(true)}>
          <Image source={icons.questionMark} style={{width: 20, height: 20}} />
        </Pressable>
      </View>

      <View style={styles.balance_view}>
        <Text>Balance</Text>
        <Image source={icons.diamond} style={styles.diamonds_img} />
        <Text>{my_data?.wallet}</Text>
      </View>
      <InfoModal
        show_model={showInfo}
        setShowModel={setShowInfo}
        faqData={faqData}
        title="Recharge Method Info"
      />
    </>
  );
};

export default RenderRechargeSheetHeader;

const styles = StyleSheet.create({
  upper_container: {
    width: width,
    alignItems: 'center',
    justifyContent: 'center',
    borderBottomWidth: 0.5,
    padding: 15,
    flexDirection: 'row',
  },
  recharge_text: {
    color: 'black',
    fontWeight: '600',
    fontSize: 16,
  },
  question_mark_view: {
    position: 'absolute',
    right: 15,
  },
  balance_view: {
    flexDirection: 'row',
    alignItems: 'center',
    marginHorizontal: 20,
    marginVertical: 10,
  },
  diamonds_img: {
    width: 15,
    height: 15,
    marginLeft: 10,
    marginRight: 3,
  },
});
