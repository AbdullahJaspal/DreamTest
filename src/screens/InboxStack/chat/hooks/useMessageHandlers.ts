import {useEffect, useCallback} from 'react';
import {Alert} from 'react-native';
import {useTranslation} from 'react-i18next';
import Toast from 'react-native-simple-toast';
import {useSocket} from '../../../../socket/SocketProvider';
import {CHAT_EVENTS} from '../../../../socket/events';

import * as userApi from '../../../../apis/userApi';
import {
  MessageStatus,
  ExtendedMessage,
  generateUniqueId,
} from '../../../../utils/chatUtils';

import {useAppSelector, useAppDispatch} from '../../../../store/hooks';
import {
  initRoomMessages,
  setMessages,
  appendMessage,
  appendMessages,
  updateMessageStatus,
  removeMessage,
  setLoadingEarlier,
  setHasMoreMessages,
  incrementPage,
  updateMessage,
} from '../../../../store/slices/content/messageSlice';

import {
  makeSelectMessagesForRoom,
  makeSelectLoadingEarlierForRoom,
  makeSelectHasMoreMessagesForRoom,
  makeSelectPageForRoom,
} from '../../../../store/selectors';

interface UserData {
  id: string;
  nickname: string;
  profile_pic: string;
  auth_token?: string;
}

export const useMessageHandlers = (
  roomId: string,
  myData: UserData,
  userData: UserData,
) => {
  const {t} = useTranslation();
  const {isConnected, emit, on} = useSocket();
  const dispatch = useAppDispatch();

  const selectMessagesForRoom = makeSelectMessagesForRoom();
  const selectLoadingEarlierForRoom = makeSelectLoadingEarlierForRoom();
  const selectHasMoreMessagesForRoom = makeSelectHasMoreMessagesForRoom();
  const selectPageForRoom = makeSelectPageForRoom();

  const messages = useAppSelector(state =>
    selectMessagesForRoom(state, roomId),
  );
  const loadingEarlier = useAppSelector(state =>
    selectLoadingEarlierForRoom(state, roomId),
  );
  const hasMoreMessages = useAppSelector(state =>
    selectHasMoreMessagesForRoom(state, roomId),
  );
  const page = useAppSelector(state => selectPageForRoom(state, roomId));

  useEffect(() => {
    dispatch(initRoomMessages({roomId}));
  }, [dispatch, roomId]);

  const prepareUserData = useCallback(
    () => ({
      _id: myData.id,
      name: myData.nickname,
      avatar: myData.profile_pic,
    }),
    [myData],
  );

  useEffect(() => {
    if (!isConnected) return;

    const handleIncomingMessage = (data: any) => {
      try {
        const formattedMessage = {
          _id: data._id || generateUniqueId(),
          senderId: data.senderId,
          receiverId: data.receiverId,
          text: data.text || '',
          type: data.type || 'text',
          image: data.image === null ? '' : data.image || '',
          video: data.video === null ? '' : data.video || '',
          document: data.document === null ? '' : data.document || '',
          audio: data.audio === null ? '' : data.audio || '',
          createdAt: data.createdAt ? new Date(data.createdAt) : new Date(),
          user: {
            _id: data.senderId,
            name:
              data.senderId === myData.id ? myData.nickname : userData.nickname,
            avatar:
              data.senderId === myData.id
                ? myData.profile_pic
                : userData.profile_pic,
          },
          roomId: roomId,
          status:
            data.senderId === myData.id
              ? MessageStatus.SENT
              : MessageStatus.RECEIVED,
        };

        console.log('My messages', messages);
        console.log('Sockets: Received message:', formattedMessage);
        console.log('Message status type:', typeof formattedMessage.status);

        const isDuplicate = messages.some(msg => msg._id === data._id);

        if (!isDuplicate) {
          console.log('Not a duplicate - appending new message');
          dispatch(appendMessage({roomId, message: formattedMessage}));
        } else {
          console.log("It's duplicated - replacing message entirely");

          // Create a completely new array with updated messages
          const updatedMessages = messages.map(msg => {
            if (msg._id === data._id) {
              return {
                ...msg,
                // Make sure to use the actual enum value
                status:
                  data.senderId === myData.id
                    ? MessageStatus.SENT
                    : MessageStatus.RECEIVED,
                image: formattedMessage.image,
                video: formattedMessage.video,
                audio: formattedMessage.audio,
                document: formattedMessage.document,
                uploadProgress: 100,
              };
            }
            return msg;
          });

          dispatch(setMessages({roomId, messages: updatedMessages}));
        }

        setTimeout(() => {
          refreshMessagesFromServer();
        }, 300);
      } catch (error) {
        console.error('Error handling incoming message:', error);
      }
    };

    // Create this as a separate function outside of the handler
    const refreshMessagesFromServer = async () => {
      try {
        console.log('Refreshing messages from server (silently)');
        const result = await userApi.getAllMessages(
          userData.id,
          myData.auth_token,
          1,
          20,
        );

        if (result?.messages?.length > 0) {
          const formattedMessages = result.messages.map(msg => ({
            ...msg,
            status: msg.isRead
              ? MessageStatus.READ
              : msg.isReceived
              ? MessageStatus.DELIVERED
              : MessageStatus.SENT,
          }));

          // Replace messages without showing loading indicator
          dispatch(setMessages({roomId, messages: formattedMessages}));

          // Update hasMoreMessages flag
          dispatch(
            setHasMoreMessages({
              roomId,
              hasMore: result.hasMore || result.messages.length >= 20,
            }),
          );
        }
      } catch (error) {
        console.error('Error silently refreshing messages:', error);
        // Don't show an error toast to avoid annoying the user
      }
    };
    const handleMessageDeleted = (data: any) => {
      dispatch(removeMessage({roomId, messageId: data.messageId}));
    };

    // Set up event listeners
    const unsubscribeMessageReceived = on(
      'customEventResponse',
      handleIncomingMessage,
    );
    const unsubscribeMessageDeleted = on(
      CHAT_EVENTS.MESSAGE_DELETED,
      handleMessageDeleted,
    );

    // Cleanup on unmount
    return () => {
      unsubscribeMessageReceived();
      unsubscribeMessageDeleted();
    };
  }, [isConnected, roomId, myData, userData, dispatch, on, messages]);

  // Fetch chat history
  const fetchChat = useCallback(
    async (pageNum = 1, append = false) => {
      try {
        if (pageNum === 1 && !append) {
          // Show loading indicator
          dispatch(
            setMessages({
              roomId,
              messages: [
                {
                  _id: 'loading',
                  text: '',
                  createdAt: new Date(),
                  user: {_id: 'system'},
                  system: true,
                } as unknown as ExtendedMessage,
              ],
            }),
          );
        }

        if (pageNum > 1) {
          dispatch(setLoadingEarlier({roomId, loading: true}));
        }

        const result = await userApi.getAllMessages(
          userData.id,
          myData.auth_token,
          pageNum,
          20,
        );

        if (result?.messages?.length > 0) {
          const formattedMessages = result.messages.map(msg => ({
            ...msg,
            status: msg.isRead
              ? MessageStatus.READ
              : msg.isReceived
              ? MessageStatus.DELIVERED
              : MessageStatus.SENT,
          }));

          if (append) {
            // For pagination, append to existing messages
            dispatch(appendMessages({roomId, messages: formattedMessages}));
          } else {
            // Initial load
            dispatch(setMessages({roomId, messages: formattedMessages}));
          }

          // Check if there are more messages to load
          dispatch(
            setHasMoreMessages({
              roomId,
              hasMore: result.hasMore || result.messages.length >= 20,
            }),
          );
        } else {
          // No messages or empty chat
          if (pageNum === 1) {
            dispatch(setMessages({roomId, messages: []}));
          }
          dispatch(setHasMoreMessages({roomId, hasMore: false}));
        }
      } catch (error) {
        console.error('Error loading messages:', error);
        Toast.show(t('Failed to load messages'), Toast.SHORT);
        if (pageNum === 1) {
          dispatch(setMessages({roomId, messages: []}));
        }
      } finally {
        dispatch(setLoadingEarlier({roomId, loading: false}));
      }
    },
    [userData.id, myData.auth_token, t, dispatch, roomId],
  );

  // Load initial messages
  useEffect(() => {
    fetchChat();
  }, [fetchChat]);

  // Load more messages (pagination)
  const handleLoadEarlier = useCallback(async () => {
    if (loadingEarlier || !hasMoreMessages) return;

    dispatch(incrementPage({roomId}));
    const nextPage = page + 1;
    await fetchChat(nextPage, true);
  }, [fetchChat, page, loadingEarlier, hasMoreMessages, dispatch, roomId]);

  const sendMessage = useCallback(
    (message: ExtendedMessage): boolean => {
      if (!isConnected) {
        console.log('Cannot send message: Socket not connected');
        return false;
      }

      try {
        const enhancedMessage = {
          ...message,
          createdAt: message.createdAt || new Date(),
          status: MessageStatus.SENDING,
          user: {
            _id: message.senderId,
            name: message.user?.name || myData.nickname,
            avatar: message.user?.avatar || myData.profile_pic,
          },
        };

        emit('customEvent', [enhancedMessage]);

        setTimeout(() => {
          dispatch(
            updateMessageStatus({
              roomId,
              messageId: message._id,
              status: MessageStatus.SENT,
            }),
          );
        }, 500);

        return true;
      } catch (error) {
        console.error('Error sending message:', error);
        dispatch(
          updateMessageStatus({
            roomId,
            messageId: message._id,
            status: MessageStatus.FAILED,
          }),
        );
        return false;
      }
    },
    [isConnected, emit, myData, dispatch, roomId],
  );

  const handleSendTextMessage = useCallback(
    newMessages => {
      const messagesToSend = newMessages.map(msg => {
        const newId = msg._id;
        return {
          _id: newId,
          senderId: myData.id,
          receiverId: userData.id,
          text: msg.text || '',
          type: 'text',
          image: '',
          video: '',
          document: '',
          audio: '',
          parentMessageId: msg.replyToMessage?._id,
          parentMessage: msg.replyToMessage,
          roomId: roomId,
          createdAt: msg.createdAt || new Date(),
          user: prepareUserData(),
          status: isConnected ? MessageStatus.SENDING : MessageStatus.FAILED,
        };
      });

      messagesToSend.forEach(message => {
        dispatch(appendMessage({roomId, message}));
      });

      if (isConnected) {
        messagesToSend.forEach(message => {
          sendMessage(message);
        });
      }
    },
    [
      isConnected,
      myData.id,
      userData.id,
      roomId,
      prepareUserData,
      sendMessage,
      dispatch,
    ],
  );

  const handleRetryMessage = useCallback(
    async (messageId: string) => {
      const messageToRetry = messages.find(msg => msg._id === messageId);
      if (!messageToRetry) return;

      dispatch(
        updateMessageStatus({
          roomId,
          messageId,
          status: MessageStatus.SENDING,
        }),
      );

      try {
        if (!isConnected) {
          throw new Error('No connection');
        }

        const success = sendMessage(messageToRetry);

        if (!success) {
          throw new Error('Failed to send message');
        }
      } catch (error) {
        dispatch(
          updateMessageStatus({
            roomId,
            messageId,
            status: MessageStatus.FAILED,
          }),
        );

        let errorMessage = t('Failed to send message. Please try again.');
        if (error instanceof Error) {
          if (error.message.includes('No connection')) {
            errorMessage = t('No internet connection. Please try again later.');
          }
        }

        Toast.show(errorMessage, Toast.SHORT);
      }
    },
    [messages, isConnected, sendMessage, dispatch, roomId, t],
  );

  const handleDeleteMessage = useCallback(
    (messageId: string) => {
      Alert.alert(
        t('Delete Message'),
        t('Are you sure you want to delete this message?'),
        [
          {text: t('Cancel'), style: 'cancel'},
          {
            text: t('Delete'),
            style: 'destructive',
            onPress: () => {
              dispatch(removeMessage({roomId, messageId}));

              if (isConnected) {
                emit(CHAT_EVENTS.MESSAGE_DELETED, {
                  messageId: messageId,
                  roomId: roomId,
                });
              }
            },
          },
        ],
      );
    },
    [isConnected, roomId, t, emit, dispatch],
  );

  const handleMessageReply = useCallback(
    (message: ExtendedMessage) => {
      Toast.show(t('Reply functionality coming soon'), Toast.SHORT);
    },
    [t],
  );

  const handleMessageForward = useCallback(
    (message: ExtendedMessage) => {
      Toast.show(t('Forward functionality coming soon'), Toast.SHORT);
    },
    [t],
  );

  return {
    messages,
    loadEarlier: loadingEarlier,
    loadingEarlier,
    hasMoreMessages,
    handleLoadEarlier,
    handleSendTextMessage,
    handleRetryMessage,
    handleDeleteMessage,
    handleMessageReply,
    handleMessageForward,
    sendMessage,
  };
};
