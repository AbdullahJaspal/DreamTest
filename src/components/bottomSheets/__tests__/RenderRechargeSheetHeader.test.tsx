import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import RenderRechargeSheetHeader from '../RenderRechargeSheetHeader';
import configureStore from 'redux-mock-store';
import {useAppSelector} from '../../../store/hooks';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useAppSelector: jest.fn(),
}));

describe('RenderRechargeSheetHeader', () => {
  const mockStore = configureStore([]);
  let store: any;

  beforeEach(() => {
    store = mockStore({
      my_data: {
        my_profile_data: {
          wallet: 1000,
        },
      },
    });

    (useAppSelector as unknown as jest.Mock).mockImplementation(selector =>
      selector(store.getState()),
    );
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <RenderRechargeSheetHeader />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
