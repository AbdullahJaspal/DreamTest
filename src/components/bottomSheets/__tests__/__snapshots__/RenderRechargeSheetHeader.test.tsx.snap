// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderRechargeSheetHeader renders correctly and matches snapshot 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 15,
        "width": 750,
      }
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Recharge
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "position": "absolute",
          "right": 15,
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/images/question_mark.png",
          }
        }
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
      />
    </View>
  </View>,
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginHorizontal": 20,
        "marginVertical": 10,
      }
    }
  >
    <Text>
      Balance
    </Text>
    <Image
      source={
        {
          "testUri": "../../../src/assets/images/diamond.png",
        }
      }
      style={
        {
          "height": 15,
          "marginLeft": 10,
          "marginRight": 3,
          "width": 15,
        }
      }
    />
    <Text>
      1000
    </Text>
  </View>,
]
`;
